File: Fase-3/EDII/tiro-ao-alvo.py
import math
def search(vR, b, e, n):
    if e==b:
        if n > vR[e]: return e+1
        else: return e
    mid = b + (e-b)//2;
    if vR[mid] == n: return mid
    elif vR[mid] > n: return search(vR, b, mid-1, n)
    else: return search(vR, mid+1, e, n)

raios = list()
r, t = map(int, input().split(' '))
p = 0
for i in range(r):
    rVal = int(input())
    raios.append(rVal)

for i in range(t):
    x, y = map(int, input().split(' '))
    d = math.sqrt(x**2 + y**2)
    p += r - search(raios, 0, r-1, d)
    #print("Distance = {}".format(d))
    #print("Index = {}".format(p))
print(p)
#print(recursiveBinarySearch(aList,455, 0, len(aList)))

File: Fase-3/EDII/binarySearchProblem/tiro-ao-alvo.cc
#include <iostream>
#include <cmath>

using namespace std;

int search(int vR[], int b, int e, float n){
  if (b == e) return (n > vR[e] ? e+1 : e);
  int m = (b+e)/2;
  if (vR[m] == n) return m;
  if (n > vR[m]) return search(vR, m+1, e, n);
  return search(vR, b, m-1, n);
}

int main() {
  int r, s, x, y, rVal, p = 0;
  float d = 0.0;

  cin >> r >> s;

  int vR[r];

  for (int i = 0; i < r; i++){
    scanf("%d", &rVal);
    vR[i] = rVal;
  }

  for (int i = 0; i < s; i++){
    scanf("%d %d", &x, &y);
    d = sqrt(x*x + y*y);
    p += search(vR, 0, r-1, d);
  }

  cout <<  (r*s) - p << endl;

  return 0;
}
